Add support for experimental docs command

Rapptz opened the pull request:
This works okay on my end. I think this could be improved but for now this works for the basic testing I did.

Daggy1234 reviewed (approved):
Looks good to me. I'll wait for @rchl and @rwols thoughts as well!

rchl commented on plugin.py:
```diff
@@ -90,6 +93,18 @@ def install_or_update(cls) -> None:
+    session_name = "rust-analyzer"
+
+    def run(self, edit: sublime.Edit) -> None:
+        params = text_document_position_params(self.view, self.view.sel()[0].b)
+        session = self.session_by_name(self.session_name)
+        session.send_request(Request('experimental/externalDocs', params), self.on_result)
+
+    def on_result(self, url: Optional[str]) -> None:
+        if url is not None:
+            self.view.window().run_command("open_url", { "url": url })
```
While not likely to be a big issue, various things should be "null-checked" for the code to be really fool proof.

 - The `sel()` can be empty
 - The `session` can be `None`
 - The `window()` can be `None`

rchl commented on plugin.py:
```diff
@@ -1,6 +1,9 @@
-from LSP.plugin import AbstractPlugin
+from LSP.plugin.core.typing import Optional
+from LSP.plugin import AbstractPlugin, Request
 from LSP.plugin import register_plugin
 from LSP.plugin import unregister_plugin
+from LSP.plugin.core.registry import LspTextCommand
+from LSP.plugin.core.views import text_document_position_params
```
Just a style nit but we typically:
 - sort the imports
 - have one import per line (apart from `typing` stuff). Kinda arbitrary choice but it became the convention. :)

rchl reviewed (commented):
There is close to 0% chance that someone will discover this functionality on its own. I think it should be added to the readme.

Rapptz commented:
I've fixed the review nits, I think.

I moved the checks over to an `is_enabled` block. I didn't know how you wanted imports sorted though so I went for alphabetical descending.

rchl commented:
> I moved the checks over to an `is_enabled` block. I didn't know how you wanted imports sorted though so I went for alphabetical ascending.

Yes, alphabetical is good.

> I will add that one line per import is weird, and the project that I was looking at [didn't use that code style](https://github.com/sublimelsp/LSP-julia/blob/main/plugin.py?rgh-link-date=2021-06-12T18%3A29%3A01Z) so I wasn't sure it existed.

A matter of getting used to. :)
As far as LSP-julia, it's not a project made by "core" members.
The LSP itself is an example of the "official" style. But it's just a style nit and I wouldn't hold it against the PR. :)

rchl commented on plugin.py:
```diff
@@ -90,6 +94,32 @@ def install_or_update(cls) -> None:
+class RustAnalyzerOpenDocsCommand(LspTextCommand):
+    session_name = "rust-analyzer"
+
+    def is_enabled(self) -> bool:
+        if not bool(self.session_by_name(self.session_name)):
+            return False
+
+        selection = self.view.sel()
+        if len(selection) == 0:
+            return False
+
+        return True
```
would suggest to call the default implementation and add that selection check only, to avoid repeating the logic.

The original implementation does the session check already.
https://github.com/sublimelsp/LSP/blob/34e9fbe52460e68aba7495cf33c33528f69f63a6/plugin/core/registry.py#L54-L69

Rapptz commented:
> There is close to 0% chance that someone will discover this functionality on its own. I think it should be added to the readme.

There is no documentation for this one in VSCode, it's only available via a default keybinding and the command palette there. I figured similar treatment was sufficient. I didn't add a keybinding because it's kind of an anti-practice in Sublime Text plugins to do so.

I'm willing to document it though, just not sure where you'd prefer.

rchl commented on plugin.py:
```diff
@@ -90,6 +94,32 @@ def install_or_update(cls) -> None:
+
+    def run(self, edit: sublime.Edit) -> None:
+        params = text_document_position_params(self.view, self.view.sel()[0].b)
+        session = self.session_by_name(self.session_name)
+        session.send_request(Request("experimental/externalDocs", params), self.on_result)
```
When I meant that session can be None, I meant more from type-checking point of view so checking verifying in `is_enabled` won't really help with that.

And yes, it's a hypothetical issue but types are not forgiving :)

rchl commented:
> I'm willing to document it though, just not sure where you'd prefer.

rust-analyzer has it documented: https://github.com/rust-analyzer/rust-analyzer/blob/a9a978e0bb418b540bd4ede0cc5a16bfb1c79baf/docs/dev/lsp-extensions.md#open-external-documentation

Maybe something like:

```
## Custom Command Palette Commands

### LSP-rust-analyzer: Open Docs Under Cursor

Opens the URL to documentation for the symbol under the cursor, if available.

Rapptz commented:
That's documentation meant for plugin writers, not consumers of the language server (it's also what I used :>). I've pushed the docs change. Thanks for the wording!
